Jenkinsfile

```diff
@@ -193,6 +193,29 @@ pipeline {
                 }
             }
         }
+        stage("Make-doxygen") {
+            agent any
+            steps {
+              if (isBranch('master')) {
+                withCredentials([usernamePassword(credentialsId: 'a630aebc-6861-4e69-b497-fd7f496ec46b',
+                    usernameVariable: 'GIT_USERNAME', passwordVariable: 'GIT_PASSWORD')]) {
+                    sh """#!/bin/bash
+                        set -x
+                        make doxygen
+                        git config --global user.email "mc.stanislaw@gmail.com"
+                        git config --global user.name "Stan Jenkins"
+                        git checkout --detach
+                        git branch -D gh-pages
+                        git checkout --orphan gh-pages
+                        git add -f ./doc/
+                        git commit -m "auto generated docs from Jenkins"
+                        git push https://${GIT_USERNAME}:${GIT_PASSWORD}@github.com/${fork()}/math.git +gh-pages:gh-pages
+                        exit 1
```

Comment:
@seantalts 

1. I'm not sure how to verify this works in Jenkins, should I switch the branch it tests for to this one first?
2. Should I also have this script remove the stan/test folders to reduce the time it takes to upload the `gh-pages` branch?
3. Does the branch need to be moved back to the original branch being tested on Jenkins after this runs?

Reply:
1. Yeah, that's probably the easiest way to do it.
2. Seems fine to rm stan test etc
3. I would add a `post { always { deleteDir() } }` to be safe and make sure things are cleaned up.
Extras:
1. You probably also need to do a `checkout scm` up front
2. You should be able to keep this as a declarative stage, I think, by doing `when { branch 'master' } ` instead of `isBranch('master')` (see https://github.com/rok-cesnovar/math/blob/9cb9a1f7aa3c46b3a5c311e14f77a0d80ddcc3b2/Jenkinsfile#L59)
3. I notice some of the git commands here are different from the ones for the clang-format step - curious if you found a better way to do stuff? I'm not really sure what the commands do differently.
4. for the gh-pages branch thing, all you have to do is push to that branch and it finds any HTML? Or do you have to move the doc somewhere specific?

Reply:
(1), (2), will do

3. I wouldn't say it's a better way to do stuff. At my work this is what I use which I stole from a guy who is pretty smart. I can make this look more like how we do `clang-format`.
4. I believe someone with admin privileges needs to go in to the settings to tell github  to look for docs in the gh-pages branch and that they are in the docs folder. Though Dan said [here](https://github.com/stan-dev/math/issues/781#issuecomment-370018091) we had this set up before so anything in gh-pages will show up at http://mc-stan.org/math. If it doesn't work then we would have to move the `api` folder to the top level or something like that

Reply:
3. I would just say they should look similar and maybe even share code (abstract into function that pushes to given branch or something?). But I'm not tied to the commands I wrote for clang-format - it mostly just worked and I didn't see a reason to iterate further. Happy to use your version too - he probably thought about it more than I did, haha.
4. Cool. I can look into this when it's ready to test (is it ready to test that part?)

Reply:
Let me play around with (3) and I'll ping you when it's ready to be looked at